endpoints/openrtb2/auction.go

```diff
@@ -1566,6 +1576,51 @@ func getJsonSyntaxError(testJSON []byte) (bool, string) {
 	return false, ""
 }
 
+func (deps *endpointDeps) processStoredAuctionResponses(ctx context.Context, requestJson []byte) (map[string]json.RawMessage, []error) {
+	// Input request may have stored request or stored imps that will modify imps from incoming request
+	// Cases:
+	// - there is only stored request id in incoming request
+	// - there is at least one stored imp with stored auction response
+	// Need to update ImpExtPrebid after stored responses/impressions applied to incoming request
+	impInfo, errs := parseImpInfo(requestJson)
+	if len(errs) > 0 {
+		return nil, errs
+	}
+
+	storedAuctionResponseIds := make([]string, 0, 0) //all stored responses ids from all imps
+	//because of bulk fetch responses we need to map imp id to stored resp body
+	impIdToRespId := make(map[string]string)              //imp id to stored resp id
+	impIdToStoredResp := make(map[string]json.RawMessage) //imp id to stored resp body
+	for index, impData := range impInfo {
+
+		if impData.ImpExtPrebid.StoredAuctionResponse != nil {
+			if len(impData.ImpExtPrebid.StoredAuctionResponse.ID) == 0 {
+				return nil, []error{fmt.Errorf("request.imp[%d] has ext.prebid.storedauctionresponse specified, but \"id\" field is missing ", index)}
+			}
+			storedAuctionResponseIds = append(storedAuctionResponseIds, impData.ImpExtPrebid.StoredAuctionResponse.ID)
+
+			impId, err := jsonparser.GetString(impData.Imp, "id")
+			if err != nil {
+				return nil, []error{err}
+			}
+
+			impIdToRespId[impId] = impData.ImpExtPrebid.StoredAuctionResponse.ID
```

Comment:
Imp ids must be unique per request, the situation you described is not possible. 
It can be the situation where 2 or more imps have same stored response:
imp-id-1 --> resp-1
imp-id-2 --> resp-1
but in this case fetcher handles this properly.

Reply:
Gotcha. I see that imp IDs are guaranteed to be unique at this point because we ensured this during validation in `endpoints/openrtb2/auction.go#validateRequest`.